pulled out method which recursively returns all keys of an object

webark opened the pull request:
Now more independent and testable. Switched to using Object.keys.

fixes #225

webark commented:
@nucleartide and/or @BillyRayPreachersSon Let me know if there are other test cases that you'd like to see.

nucleartide commented:
Nice! Let's also add a test for `Object.create(null)`, which doesn't have any prototype methods. 🙂

webark commented:
@nucleartide added in that test, and fixed the ember object test. Also switched it from a forEach to a reduce. Seems a lot simpler now. Because it was basically rewritten, I force pushed over that commit. Let me know if you have any other edgecases you'd like addressed.

nucleartide commented on addon/index.js:
```diff
@@ -4,6 +4,7 @@ import objectToArray from 'ember-changeset/utils/computed/object-to-array';
 import isEmptyObject from 'ember-changeset/utils/computed/is-empty-object';
 import isPromise from 'ember-changeset/utils/is-promise';
 import isObject from 'ember-changeset/utils/is-object';
+import recursiveKeys from 'ember-changeset/utils/recursive-keys';
```
Let's change the name of this util to `leafKeys`, since this util excludes intermediary keys.

nucleartide commented on addon/index.js:
```diff
@@ -201,7 +202,7 @@ export function changeset(obj, validateFn = defaultValidatorFn, validationMap =
 
-        this._allKeys(resetProperties).forEach(function(key) {
+        recursiveKeys(resetProperties).forEach(function(key) {
           if (!changedKeys.includes(key)) {
```
I missed this, but let's wrap `changedKeys` with `Ember.A` here to ensure the `.includes` method exists.

nucleartide replied:
(Unsure what the new Ember import would be here.)

nucleartide commented on addon/utils/recursive-keys.js:
```diff
@@ -0,0 +1,48 @@
+import { isArray } from '@ember/array';
+
+/**
+ * Recursivly loops over all sub object and
```
Let's add something about how this function only returns child-most (or leaf-most) keys. Example:

```js
/**
 * Given an object, return an array of the leaf-most keys
 * in nested string format.
 */
```

nucleartide commented on addon/utils/recursive-keys.js:
```diff
@@ -0,0 +1,48 @@
+ * Recursivly loops over all sub object and
+ * returns an array of keys in nesteted string format.
+ *
+ * eg. this object
```
Let's remove this example, I'd rather have people look at the tests.

nucleartide commented on addon/utils/recursive-keys.js:
```diff
@@ -0,0 +1,48 @@
+ * ]
+ *
+ * @param  {Object} [object={}] Object whos keys to iterate over.
+ * @param  {String} [scope=''] A sring that can be prepended to the key.
```
Nitpick: change to `A string that can be prepended to the object's keys.`

nucleartide commented on addon/utils/recursive-keys.js:
```diff
@@ -0,0 +1,48 @@
+ *   'item.innerItem.nest.other',
+ * ]
+ *
+ * @param  {Object} [object={}] Object whos keys to iterate over.
```
Nitpick: change to `Object to inspect.`

nucleartide commented on addon/utils/recursive-keys.js:
```diff
@@ -0,0 +1,48 @@
+ * @param  {String} [scope=''] A sring that can be prepended to the key.
+ * @return {Array}
+ */
+const recursiveKeys = function(object = {}, scope = '') {
```
Let's make this `scope` param private. So:

```js
export default function leafKeys(object = {}) {
  return helper(object);
}

function helper(object = {}, scope = '') {
  // ...
}
```

nucleartide commented on tests/unit/utils/recursive-keys-test.js:
```diff
@@ -0,0 +1,152 @@
+
+module('Unit | Utility | recursive-keys');
+
+test('it returns an array with keys as nested strings', function(assert) {
```
Since we're repeating the same assertion, let's write these tests in table-driven form. Example:

```js
let testData = [
  {
    desc: 'it returns an array with keys as nested strings',
    input: {
      topValue: -1,
      item: {
        secondValue: 0,
        innerItem: {
        thing: 1,
        other: 2,
        nest: {
          thing: 3,
          other: 4
        }
      }
    },
    expectedResult: [
      'topValue',
      'item.secondValue',
      'item.innerItem.thing',
      'item.innerItem.other',
      'item.innerItem.nest.thing',
      'item.innerItem.nest.other',
    ]
  },

  // ...
];

testData.forEach(({ desc, input, expected }) => {
  test(desc, function(assert) {
    assert.deepEqual(leafKeys(input), expected, desc);
  });
});
```

nucleartide commented on tests/unit/utils/recursive-keys-test.js:
```diff
@@ -0,0 +1,152 @@
+        thing: 1,
+        other: 2,
+        empty: {},
+        nest: Object.create(null),
```
Let's add a test case where `Object.create(null)` isn't at the "leaf" level. So for example:

```js
let o = Object.create(null);
o.foo = 'bar';
o.hello = 'world';
o.someObj = { blah: 'blah' };

nestedObject.item = o;
```

BillyRayPreachersSon commented:
Thanks both for the quick turnaround, this looks like it should fix my problem. Once this is released, I'll update my app and let you know how it goes.

webark commented:
@nucleartide I think I covered all of your comments. Let me know if you have any more.

nucleartide commented:
@webark Looks amazing! Thanks, going to merge this.

sfanetti commented:
I have a model that has a belongsTo relationship with another DS model.  The DS model - has a cyclical relationship within it because of an included service.  

This caused the stack to blow out with the previously implemented release.  Though I have not yet tested this - I suspect it will also have the same issue.  Any models that happen to internal circular relationships will get blown out with this. 

It would be good to have a schema of properties to set when executing the changeset instead of doing the recursive scan.  This limits the work done in the method and provides better control over what gets copied from the changeset to the saving one.  Do you have any suggestions about a strategy for approaching this within this library?

Maybe at construction - options can be passed in with a structure that outlines the relevant attributes that are mutable within the form so that nesting depth can be controlled?

nucleartide commented:
@sfanetti Please see my comment at https://github.com/poteto/ember-changeset/issues/239#issuecomment-350291579.

BillyRayPreachersSon commented:
I've just tested the latest code and I seem to have a slightly different stack issue than @sfanetti (based on @nucleartide's comment in https://github.com/poteto/ember-changeset/issues/239).

I've put the details into https://github.com/poteto/ember-changeset/issues/244.
